Add some doc for export_for_training

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135918
* #135912
* #135080

Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

tugsbayasgalan committed:
Add some doc for export_for_training

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

avikchaudhuri commented on docs/source/export.rst:
```diff
@@ -291,6 +291,176 @@ torch.export>`), but seeing as the context manager does not affect the tensor
 
+.. _Training Export:
+
+Training Export
```
Not a good title. Maybe something like "Export for Training and Inference"?

avikchaudhuri commented on docs/source/export.rst:
```diff
@@ -291,6 +291,176 @@ torch.export>`), but seeing as the context manager does not affect the tensor
+
+In this API, we produce the most generic IR that contains all ATen operators
+(including both functional and non-functional) which can be used to train in
+eager PyTorch Autograd. This API is intended for PT2 quantization training use cases
```
I'm not sure why you need to mention quantization training. E.g., you'd want this for normal eager training too.

avikchaudhuri commented on docs/source/export.rst:
```diff
@@ -291,6 +291,176 @@ torch.export>`), but seeing as the context manager does not affect the tensor
+In this API, we produce the most generic IR that contains all ATen operators
+(including both functional and non-functional) which can be used to train in
+eager PyTorch Autograd. This API is intended for PT2 quantization training use cases
+and will soon be the default IR of torch.export.export in the near future. To read further about
```
soon vs. near future, pick one

avikchaudhuri commented on docs/source/export.rst:
```diff
@@ -291,6 +291,176 @@ torch.export>`), but seeing as the context manager does not affect the tensor
+    )
+    Range constraints: {}
+
+Here you can see that, we kept `conv2d` op in the IR while decomposing the rest. Now the IR is an functional IR
```
Edited:

Here you can see that we kept `conv2d` op in the IR while decomposing the rest. Now the IR is a functional IR
containing core aten operators except for `conv2d`.

You can do even more customization by directly registering your chosen decomposition behaviors.

avikchaudhuri commented on docs/source/export.rst:
```diff
@@ -291,6 +291,176 @@ torch.export>`), but seeing as the context manager does not affect the tensor
+as :func:`export` except for the operators in the graph. You can see that we captured `batch_norm` in the most general
+form. This op is non-functional and will be lowered to different ops when running under inference mode.
+
+You can also go from this IR to an inference IR via :func:`run_decompositions` with arbitrary customizations
```
Edited:

From the above output, you can see that :func:`export_for_training` produces pretty much the same `ExportedProgram`
as :func:`export` except for the operators in the graph. You can see that we captured `batch_norm` in the most general
form. This op is non-functional and will be lowered to different ops when running inference.

You can also go from this IR to an inference IR via :func:`run_decompositions` with arbitrary customizations.

avikchaudhuri commented on docs/source/export.rst:
```diff
@@ -291,6 +291,176 @@ torch.export>`), but seeing as the context manager does not affect the tensor
+the motivation behind this change, please refer to
+https://dev-discuss.pytorch.org/t/why-pytorch-does-not-need-a-new-standardized-operator-set/2206
+
+With this API, and :func:`run_decompositions()`, you should be able to get any inference IR with
```
Edited:

When this API is combined with :func:`run_decompositions()`, you should be able to get inference IR with
any desired decomposition behavior.

avikchaudhuri reviewed (approved):
Looks good, suggested edits for grammar / style.

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add some doc for export_for_training"


Differential Revision: [D62610491](https://our.internmc.facebook.com/intern/diff/D62610491)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62610491).
